crates/sc-proof-of-time/src/verifier.rs

```diff
@@ -90,32 +111,53 @@ impl PotVerifier {
                         async move {
                             // Result doesn't matter here
                             let _ = result_sender
-                                .send(verifier.derive_next_seed(seed, slot_iterations).await);
+                                .send(verifier.calculate_proof(seed, slot_iterations).await);
                         }
                     });
                 }
             });
 
-            seed = match result_receiver.await {
-                Ok(Some(seed)) => seed,
-                _ => {
-                    return false;
-                }
+            let Ok(Some(calculated_proof)) = result_receiver.await else {
+                return false;
             };
 
             slots -= 1;
+
+            if slots == 0 {
+                return proof == calculated_proof;
+            }
+
+            #[cfg(feature = "pot")]
+            {
+                slot = slot + Slot::from(1);
+            }
+
+            #[cfg(feature = "pot")]
+            if let Some(parameters_change) = maybe_parameters_change
+                && parameters_change.slot == slot
+            {
+                slot_iterations = parameters_change.slot_iterations;
+                seed = calculated_proof.seed_with_entropy(&parameters_change.entropy);
+                maybe_parameters_change.take();
+            } else {
+                seed = calculated_proof.seed();
+            }
+            #[cfg(not(feature = "pot"))]
+            {
+                seed = calculated_proof.seed();
+            }
         }
     }
 
     /// Derive next seed, proving might be used if necessary
```

Comment:
What is the condition of this "necessary"?

Reply:
If proof for this input was already verified recently it'll be cached and all that would be necessary is to retrieve it from there. But if not, then it would be necessary to do the full proving to get there.

Basically there is no guarantee we have seen the proof previously.

Reply:
Can we wait until we see it? Gossip may arrive out of order. It's strange for a "verifier" to do local evaluation or maybe I'm missing something. Is this the fallback to local evaluation case that was in older specs?

Reply:
This can be called from block verification, in which case we can't wait: we either accept block or reject. This is designed as a fallback and there is a TODO in https://github.com/subspace/subspace/pull/1947 to implement "check only" version later. That is part of the "optimizations" category I mentioned a few times already.